### PR TITLE
fix(s3-request-presigner): skip hoisting SSE headers

### PR DESCRIPTION
**Pull request by AllanZhengYP**
*Issue #, if available:*
fix: https://github.com/aws/aws-sdk-js-v3/issues/1576

*Description of changes:*
When `signature-v4` generate a presigned url, it will try to hoist the headers to query strings, so these headers will exist in the url, hense easy to use. However, S3 requres server-side encryption headers to be signed in headers due to [S3 limitation](https://github.com/aws/aws-sdk-ruby/issues/874#issuecomment-135596055). So this change solve the issue by 2 changes:
1. Add a new config to the `signature-v4` presign config `unhoistableHeaders`. The presigner won't hoist the given headers to query before presign it.
2. In `s3-request-presigner`, it detects all the server-side encryption headers, and supply them to the `unhoistableHeaders` config of the presigner.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/s3-request-presigner/README.md [hunk 1]
```diff
@@ -77,3 +81,12 @@ const signer = new S3RequestPresigner({
   ...s3.config,
 });
 ```
+
+If your request contains server-side encryption(`x-amz-server-side-encryption*`)
+headers, because of S3 limitation, you need to send these headers along
+with the presigned url. That is to say, the url only from calling `formatUrl()`
+to `presigned` is not sufficient to make a request. You need to send the
+server-side encryption headers along with url. These headers remain in the
```
Comment (vecerek) [inline packages/s3-request-presigner/README.md:89]: Nit:
```suggestion
server-side encryption headers along with the url. These headers remain in the
```

File: packages/s3-request-presigner/src/presigner.spec.ts [hunk 2]
```diff
@@ -87,4 +87,19 @@ describe("s3 presigner", () => {
...
     });
   });
+
+  it("should disable hoisting server-side-encryption headers to query", async () => {
+    const signer = new S3RequestPresigner(s3ResolvedConfig);
+    const signed = await signer.presign({
+      ...minimalRequest,
+      headers: {
+        ...minimalRequest.headers,
+        "x-amz-server-side-encryption": "kms",
```
Comment (vecerek) [inline packages/s3-request-presigner/src/presigner.spec.ts:97]: Small suggestion: also test other headers, ones prefixed with `x-amz-server-side-encryption` to make sure all cases are covered, e.g. `x-amz-server-side-encryption-aws-kms-key-id`.

File: packages/types/src/signature.ts [hunk 3]
```diff
@@ -52,6 +52,17 @@ export interface RequestPresigningArguments extends RequestSigningArguments {
...
   expiresIn?: number;
+
+  /**
+   * A set of string whose members represents headers that should not be hoisted
+   * to presigned request's query string. If not supplied, the presigner will
+   * move all the AWS-specific headers(starting with `x-amz-`) to the request
+   * query string. If supplied, these headers remain in the presigned request's
+   * header.
+   * All headers in the provided request will have their names converted to
+   * lower case and then checked for existence in the unhoistableHeaders set.
```
Comment (vecerek) [inline packages/types/src/signature.ts:63]: Nit:
```suggestion
   * A set of strings whose representing headers that should not be hoisted
   * to presigned request's query string. If not supplied, the presigner
   * moves all the AWS-specific headers (starting with `x-amz-`) to the request
   * query string. If supplied, these headers remain in the presigned request's
   * header.
   * All headers in the provided request will have their names converted to
   * lower case and then checked for existence in the unhoistableHeaders set.
```

**Review by vecerek — COMMENTED** [review]
FWIW, I think this PR is great 🙂 Thank you so much for fixing this!

**Review by alexforsyth — APPROVED** [review]
LGTM!

**Pull request merged**